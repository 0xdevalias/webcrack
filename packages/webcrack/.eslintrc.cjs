/** @type {import("eslint").Linter.Config} */
module.exports = {
  root: true,
  extends: ['@webcrack/eslint-config/index.js'],
  ignorePatterns: ['tmp/**/*', 'test/samples/**/*'],
};
