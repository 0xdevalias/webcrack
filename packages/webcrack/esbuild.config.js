import esbuild from 'esbuild';

const args = process.argv.slice(2);
const watch = args.length > 0 && /^(?:--watch|-w)$/i.test(args[0]);

/**
 * @type {esbuild.BuildOptions[]}
 */
const configs = [
  {
    entryPoints: ['src/index.ts'],
  },
  {
    entryPoints: [
      {
        in: 'src/cjs-wrapper.ts',
        out: 'index',
      },
    ],
    outExtension: { '.js': '.cjs' },
    format: 'cjs',
    bundle: false,
  },
  {
    entryPoints: ['src/cli.ts'],
    bundle: false,
  },
];

for (const config of configs) {
  const ctx = await esbuild.context({
    bundle: true,
    format: 'esm',
    platform: 'node',
    outdir: 'dist',
    sourcemap: true,
    packages: 'external',
    logLevel: 'info',
    ...config,
  });

  if (watch) {
    await ctx.watch();
  } else {
    await ctx.rebuild();
    await ctx.dispose();
  }
}
