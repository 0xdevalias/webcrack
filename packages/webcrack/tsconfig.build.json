{
  "extends": "@webcrack/typescript-config/base.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./dist",
    "emitDeclarationOnly": true
  },
  "include": [
    "src"
  ]
}
