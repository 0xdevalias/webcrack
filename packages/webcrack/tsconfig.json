{
  "extends": "@webcrack/typescript-config/base.json",
  "compilerOptions": {
    "noEmit": true
  },
  "include": [
    "src",
    "test",
  ]
}
