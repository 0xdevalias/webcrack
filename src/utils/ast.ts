import generate from '@babel/generator';
import * as t from '@babel/types';

export function codePreview(node: t.Node) {
  const { code } = generate(node, { compact: true, comments: false });
  if (code.length > 100) {
    return code.slice(0, 70) + ' â€¦ ' + code.slice(-30);
  }
  return code;
}
