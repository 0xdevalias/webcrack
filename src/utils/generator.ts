import babelGenerate, { GeneratorOptions } from '@babel/generator';
import * as t from '@babel/types';

const defaultOptions: GeneratorOptions = { jsescOption: { minimal: true } };

export function generate(
  ast: t.Node,
  options: GeneratorOptions = defaultOptions
): string {
  return babelGenerate(ast, options).code;
}
