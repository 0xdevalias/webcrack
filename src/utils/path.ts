import { dirname, relative } from 'node:path';

export function relativePath(from: string, to: string) {
  const relativePath = relative(dirname(from), to);
  return relativePath.startsWith('.') ? relativePath : './' + relativePath;
}
