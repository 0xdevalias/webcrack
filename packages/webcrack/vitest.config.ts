import { join } from 'node:path';
import { defineProject } from 'vitest/config';

export default defineProject({
  test: {
    root: join(__dirname, 'test'),
    setupFiles: 'setup.ts',
    include: ['**/*.test.ts'],
    isolate: false,
  },
});
