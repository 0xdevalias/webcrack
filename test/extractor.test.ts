import assert from 'assert';
import { readFile } from 'fs/promises';
import { describe, expect, test } from 'vitest';
import webcrack from '../src/index';
import { relativePath } from '../src/utils/path';

describe('extractor', async () => {
  test('webpack array', async () => {
    const { bundle } = await webcrack(
      await readFile('./test/samples/webpack.js', 'utf8')
    );
    expect(bundle).toBeDefined();
    assert(bundle);
    expect(bundle.type).toBe('webpack');
    expect(bundle.entryId).toBe(2);
    expect(bundle.modules.size).toBe(3);
    for (const module of bundle.modules.values()) {
      expect(module.ast).toMatchSnapshot();
    }
  });

  test('webpack object', async () => {
    const { bundle } = await webcrack(
      await readFile('./test/samples/webpack_object.js', 'utf8')
    );
    expect(bundle).toBeDefined();
    assert(bundle);
    expect(bundle.type).toBe('webpack');
    expect(bundle.entryId).toBe(386);
    expect(bundle.modules.size).toBe(2);
    for (const module of bundle.modules.values()) {
      expect(module.ast).toMatchSnapshot();
    }
  });

  test('path mapping', async () => {
    const { bundle } = await webcrack(
      await readFile('./test/samples/webpack.js', 'utf8'),
      { mappings: m => ({ './utils/color.js': m.stringLiteral('#FBC02D') }) }
    );
    expect(bundle).toBeDefined();
    assert(bundle);
    for (const module of bundle.modules.values()) {
      expect(module.ast).toMatchSnapshot();
    }
  });
});

test('relative paths', () => {
  expect(relativePath('./a.js', './x/y.js')).toBe('./x/y.js');
  expect(relativePath('./x/y.js', './a.js')).toBe('../a.js');
});
