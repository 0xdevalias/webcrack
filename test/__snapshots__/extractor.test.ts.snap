// Vitest Snapshot v1

exports[`extractor > path mapping 1`] = `const a = require("./3.js");`;

exports[`extractor > path mapping 2`] = `
const a = require("../1.js");
const _module = 1;
module.exports.color = "#FBC02D";
{
  const _module2 = 2;
  console.log(_module2);
  console.log(module);
}
exports.a = 3;
`;

exports[`extractor > path mapping 3`] = `module.exports = 4;`;

exports[`extractor > webpack array 1`] = `const a = require("./3.js");`;

exports[`extractor > webpack array 2`] = `
const a = require("./1.js");
const _module = 1;
module.exports.color = "#FBC02D";
{
  const _module2 = 2;
  console.log(_module2);
  console.log(module);
}
exports.a = 3;
`;

exports[`extractor > webpack array 3`] = `module.exports = 4;`;

exports[`extractor > webpack object 1`] = `const r = require("./387.js").default;`;

exports[`extractor > webpack object 2`] = `module.exports = "test";`;
